Console exporter display exponential histogram buckets

alanwest opened the pull request:
Uses same format as the collector. For sake of comparison, I've included the output from our exporter and the collector's exporter using the following code.

### Code

```csharp
var histogram = meter.CreateHistogram<double>("histogram");
histogram.Record(-1000.01);
histogram.Record(-1000);
histogram.Record(0);
histogram.Record(1000);
histogram.Record(1000.01);
```

### .NET console exporter

```text
Export histogram, Meter: TestMeter
(2023-03-29T19:21:36.9011200Z, 2023-03-29T19:21:36.9130980Z] ExponentialHistogram
Value: Sum: 0 Count: 5 Min: -1000.01 Max: 1000.01 
Bucket [-1000.009769, -1000.010430), Count: 1
Bucket [-1000.009108, -1000.009769), Count: 0
Bucket [-1000.008447, -1000.009108), Count: 0
Bucket [-1000.007786, -1000.008447), Count: 0
Bucket [-1000.007125, -1000.007786), Count: 0
Bucket [-1000.006464, -1000.007125), Count: 0
Bucket [-1000.005803, -1000.006464), Count: 0
Bucket [-1000.005142, -1000.005803), Count: 0
Bucket [-1000.004480, -1000.005142), Count: 0
Bucket [-1000.003819, -1000.004480), Count: 0
Bucket [-1000.003158, -1000.003819), Count: 0
Bucket [-1000.002497, -1000.003158), Count: 0
Bucket [-1000.001836, -1000.002497), Count: 0
Bucket [-1000.001175, -1000.001836), Count: 0
Bucket [-1000.000514, -1000.001175), Count: 0
Bucket [-999.999853, -1000.000514), Count: 1
Bucket [0, 0], Count: 1
Bucket (999.999853, 1000.000514], Count: 1
Bucket (1000.000514, 1000.001175], Count: 0
Bucket (1000.001175, 1000.001836], Count: 0
Bucket (1000.001836, 1000.002497], Count: 0
Bucket (1000.002497, 1000.003158], Count: 0
Bucket (1000.003158, 1000.003819], Count: 0
Bucket (1000.003819, 1000.004480], Count: 0
Bucket (1000.004480, 1000.005142], Count: 0
Bucket (1000.005142, 1000.005803], Count: 0
Bucket (1000.005803, 1000.006464], Count: 0
Bucket (1000.006464, 1000.007125], Count: 0
Bucket (1000.007125, 1000.007786], Count: 0
Bucket (1000.007786, 1000.008447], Count: 0
Bucket (1000.008447, 1000.009108], Count: 0
Bucket (1000.009108, 1000.009769], Count: 0
Bucket (1000.009769, 1000.010430], Count: 1
```

### Collector logging exporter

```text
2023-03-29T19:24:40.614Z	info	ResourceMetrics #0
Resource SchemaURL: 
Resource attributes:
     -> service.name: Str(myservice)
     -> service.instance.id: Str(7a1a598d-ed31-410d-ae57-864a652e3433)
ScopeMetrics #0
ScopeMetrics SchemaURL: 
InstrumentationScope TestMeter 
Metric #0
Descriptor:
     -> Name: histogram
     -> Description: 
     -> Unit: 
     -> DataType: ExponentialHistogram
     -> AggregationTemporality: Cumulative
ExponentialHistogramDataPoints #0
StartTimestamp: 2023-03-29 19:24:40.525425 +0000 UTC
Timestamp: 2023-03-29 19:24:40.612426 +0000 UTC
Count: 5
Sum: 0.000000
Min: -1000.010000
Max: 1000.010000
Bucket (-1000.010430, -1000.009769], Count: 1
Bucket (-1000.009769, -1000.009108], Count: 0
Bucket (-1000.009108, -1000.008447], Count: 0
Bucket (-1000.008447, -1000.007786], Count: 0
Bucket (-1000.007786, -1000.007125], Count: 0
Bucket (-1000.007125, -1000.006464], Count: 0
Bucket (-1000.006464, -1000.005803], Count: 0
Bucket (-1000.005803, -1000.005142], Count: 0
Bucket (-1000.005142, -1000.004480], Count: 0
Bucket (-1000.004480, -1000.003819], Count: 0
Bucket (-1000.003819, -1000.003158], Count: 0
Bucket (-1000.003158, -1000.002497], Count: 0
Bucket (-1000.002497, -1000.001836], Count: 0
Bucket (-1000.001836, -1000.001175], Count: 0
Bucket (-1000.001175, -1000.000514], Count: 0
Bucket (-1000.000514, -999.999853], Count: 1
Bucket [0, 0], Count: 1
Bucket [999.999853, 1000.000514), Count: 1
Bucket [1000.000514, 1000.001175), Count: 0
Bucket [1000.001175, 1000.001836), Count: 0
Bucket [1000.001836, 1000.002497), Count: 0
Bucket [1000.002497, 1000.003158), Count: 0
Bucket [1000.003158, 1000.003819), Count: 0
Bucket [1000.003819, 1000.004480), Count: 0
Bucket [1000.004480, 1000.005142), Count: 0
Bucket [1000.005142, 1000.005803), Count: 0
Bucket [1000.005803, 1000.006464), Count: 0
Bucket [1000.006464, 1000.007125), Count: 0
Bucket [1000.007125, 1000.007786), Count: 0
Bucket [1000.007786, 1000.008447), Count: 0
Bucket [1000.008447, 1000.009108), Count: 0
Bucket [1000.009108, 1000.009769), Count: 0
Bucket [1000.009769, 1000.010430), Count: 1
```

reyang commented on src/OpenTelemetry.Exporter.Console/ConsoleMetricExporter.cs:
```diff
@@ -146,8 +146,41 @@ public override ExportResult Export(in Batch<Metric> batch)
+                                    negativeBucketCounts = new List<string>();
+                                }
+
+                                var lowerBound = Math.Exp(offset * factor).ToString("F6", CultureInfo.InvariantCulture);
```
Curious, where is this `F6` coming from, does it provide a good precision (rather than having two different buckets printed out the same thing)?

reyang commented on src/OpenTelemetry.Exporter.Console/ConsoleMetricExporter.cs:
```diff
@@ -146,8 +146,41 @@ public override ExportResult Export(in Batch<Metric> batch)
+
+                            if (exponentialHistogramData.ZeroCount != 0)
+                            {
+                                bucketsBuilder.AppendLine($"Bucket [0, 0], Count: {exponentialHistogramData.ZeroCount}");
```
I guess ZeroCount is for `[0, IEEE754 epsilon / 2]`.

alanwest replied:
From a learning perspective, I think the output helps folk derive an intuition for exponential histograms. I'm not opposed to getting more technical and surfacing up details of the underlying IEEE 754 representation of things, but I wonder if it's helpful.

alanwest replied:
English works for me.

reyang commented on src/OpenTelemetry.Exporter.Console/ConsoleMetricExporter.cs:
```diff
@@ -146,8 +146,41 @@ public override ExportResult Export(in Batch<Metric> batch)
+                                }
+
+                                var lowerBound = Math.Exp(offset * factor).ToString("F6", CultureInfo.InvariantCulture);
+                                var upperBound = Math.Exp(++offset * factor).ToString("F6", CultureInfo.InvariantCulture);
```
Is this going to give the accurate IEEE 754 representation or there will be drifts due to precision losses or we don't know?

alanwest replied:
I'm pretty confident this is absolutely imprecise 😆. I guess we need to consider the audience and use case we're trying to serve.

To me this verbose format is mainly for learning, debugging, non-production kind of use cases.

`F6` truncates things to 6 decimal places so that the width of the output is somewhat fixed and easier to read. I chose 6 simply because that's what the collector seems to display (might just be default formatting in Go).

alanwest replied:
I essentially implemented the `LowerBoundary` reverse mapping method here 

https://github.com/open-telemetry/opentelemetry-specification/blob/main/specification/metrics/data-model.md#all-scales-use-the-logarithm-function.

My understanding is that this is valid for both positive and negative scales. There is a note about the possibility that the maximum index is incorrectly computed as `+Inf`. I can verify.

utpilla commented:
Is this discrepancy in the outputs expected?

### ConsoleExporter
```text
Bucket [-1000.009769, -1000.010430), Count: 1
Bucket [-999.999853, -1000.000514), Count: 1
```

### Collector
```text
Bucket (-1000.010430, -1000.009769], Count: 1
Bucket (-1000.000514, -999.999853], Count: 1
```

alanwest commented:
> Is this discrepancy in the outputs expected?

Good eye. Had the bounds reversed. Fixed. You might have also noticed the difference in the interval notation. This is a bug in the collector. I fixed it yesterday https://github.com/open-telemetry/opentelemetry-collector/pull/7445
